### PR TITLE
WIP: Add new last-sync time indicator

### PR DESCRIPTION
**Pull request by dmsnell**
Following Simperium/simperium-android#229 this adds a new listener,
the `LastSyncTimeCache`, to track the sync status of notes. It tracks:

 - are the local changes that have not received confirmation from the server?
 - when was the last time changes from the server arrived?

These two pieces of information can be presented to indicate:

 - is it safe for me to quit out of the app? (`isSynced`)
 - do I have the edits I made on my other device? (`lastSyncTime`)

Components in the app may poll the cache for individual notes or subscribe
as a listener and receive updates.

## Todo

 - [ ] persist `LastSyncTimeCache` so it's available across app sessions
 - [x] listen during indexing operations so we don't have to wait for changes to occur before populating the values
 - [x] add UI in the note info panel showing last sync time
 - [ ] stop background sync process once all local changes have been acknowledged
 - [x] allow multiple listeners to the `LastSyncTimeCache` (currently only one is allowed)
 - [x] remove extra log statements

### CODE REVIEW

**Comment by theck13** [comment]
#### Preface
I understand this is a proof-of-concept at the moment.  I just wanted make a few comments on my initial thoughts.

#### Feedback
What does it mean for a note to be synced?  Does that mean it was saved remotely on the server or locally on the device?  I made some edits to a note while in airplane mode during testing.  Both the modified date and synced date updated while I was still offline.  I expected the modified date to change, but not the synced date.

The synced date information may be best in the **_Information_** bottom sheet.  See the screenshot below for illustration.

<kbd><a href="https://user-images.githubusercontent.com/3827611/92262937-1781a200-ee99-11ea-952d-74df4b67af78.png"></a></kbd>

File: Simplenote/src/main/java/com/automattic/simplenote/LastSyncTimeCache.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+        if (null != mListener) {
+            mListener.onUpdate(entityId, now);
+        }
+    }
+
+    public Bucket.Listener<T> listener = new Bucket.Listener<T>() {
+        @Override
+        public void onSaveObject(Bucket<T> bucket, T object) {
+            updateSyncTime(object.getSimperiumKey());
```
Comment (dmsnell) [inline Simplenote/src/main/java/com/automattic/simplenote/LastSyncTimeCache.java]: @theck13 this is the line I'm guessing is causing the sync time to update when offline.

originally I didn't have this here but then I wasn't getting a time update when making changes on the device. that is, we're updating when we receive updates from the server but we're not listening for the acknowledged change when the patch originated locally.

I'll do more digging and see if I can find another way to get that info

Comment (theck13) [reply Simplenote/src/main/java/com/automattic/simplenote/LastSyncTimeCache.java]: > we're updating when we receive updates from the server but we're not listening for the acknowledged change when the patch originated locally.

That's exactly right.  `onSaveObject` is triggered when an object is saved in Simperium Android _locally_.  `onNetworkChange` is triggered when an object is updated _remotely_.

**Comment by dmsnell** [comment]
Playing around with some updates: trying to figure out if we have enough info in the app to do a binary "is sync'd" indicator

**Comment by sshaikh** [comment]
If I may comment, I'd suggest a visible label rather than one hidden in a properties dialog (I've posted reasoning here #1181)

**Comment by theck13** [comment]
I added a **_Synced_** field to the **_Information_** sheet to show the last time a note was synced.  Currently, the last sync time cache is cleared after the app is closed.  So, only changes made while the app is open will be cached.  If a note does not have a last sync time cached, the **_Synced_** field in the **_Information_** sheet will be hidden.  Storing the last sync time cache so that it will persist across app launches will be added in a subsequent pull request.  See the screenshots below for illustration.

1140_last_sync_time

### Test
Since the last sync time cache can be affected by local and remote edits, using two devices to test is best.  I used an Android emulator and the web app.
1. **Android**: Tap any note in list.
2. **Web**: Tap note in list from Step 1.
3. **Android**: Tap ***Information*** action in app bar.
4. **Android**: Notice ***Synced*** field is hidden.
5. **Android**: Tap anywhere outside ***Information*** sheet to dismiss.
6. **Web**: Edit note from Step 1.
7. Wait ten seconds.
8. **Android**: Tap ***Information*** action in app bar.
9. **Android**: Notice ***Synced*** field is shown.
10. **Android**: Enable airplane mode.
11. **Android**: Edit note from Step 1.
12. **Android**: Tap ***Information*** action in app bar.
13. **Android**: Notice ***Synced*** field is shown and unchanged.
14. **Android**: Notice ***Modified*** field is shown and changed.
15. **Android**: Tap anywhere outside ***Information*** sheet to dismiss.
16. Wait one minute.
17. **Android**: Disable airplane mode.
18. Wait ten seconds.
19. **Android**: Tap ***Information*** action in app bar.
20. **Android**: Notice ***Synced*** field is shown and changed.
21. **Android**: Notice ***Modified*** field is shown and unchanged.

**Comment by theck13** [comment]
@dmsnell, the changes in this pull request look good to me.  Let me know if you have any suggestions from the commits I added.  If there are no other changes, I'll wait to merge this until next week when @SylvesterWilmott gets a chance to look at it.

**Comment by SylvesterWilmott** [comment]
Looks good! Thanks

**Review by theck13 — APPROVED** [review]

**Pull request merged**